# Data Cleaning and Preprocessing

## Overview
This notebook performs data quality checks and cleaning on the UCI Online Retail dataset.

## Data Source
- **Dataset**: Online Retail Dataset from UCI Machine Learning Repository
- **Period**: December 2010 - December 2011
- **Business**: UK-based online gift retailer (B2B and B2C)

## Cleaning Steps
1. Handle missing values (specifically CustomerID - required for customer analysis)
2. Remove duplicate transactions
3. Correct data types (dates, IDs)
4. Filter invalid records (non-product entries)
5. Document business rules for data anomalies

## Key Business Rules Discovered
- **InvoiceNo starting with 'C'**: Indicates a **cancellation/return** (negative quantities)
- **StockCode = POST, D, M, C2, etc.**: Non-product entries (postage, discounts, manual adjustments)
- **Valid StockCode pattern**: 5 digits optionally followed by letters (e.g., 85123A)

In [1]:
import pandas as pd

## 1. Load and Inspect Raw Data

In [2]:
# Load data (UTF-8, written by get_data.py)
df_raw = pd.read_csv("data.csv", encoding='utf-8')

print("=== Raw Data Overview ===")
print(f"Shape: {df_raw.shape[0]:,} rows x {df_raw.shape[1]} columns")
print(f"\nColumns: {list(df_raw.columns)}")

=== Raw Data Overview ===
Shape: 541,909 rows x 8 columns

Columns: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']


In [3]:
df_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  str    
 1   StockCode    541909 non-null  str    
 2   Description  540455 non-null  str    
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  str    
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 33.1 MB


In [4]:
df_raw.head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/01/2010 08:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/01/2010 08:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/01/2010 08:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/01/2010 08:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/01/2010 08:26,3.39,17850.0,United Kingdom
5,536365,22752,SET 7 BABUSHKA NESTING BOXES,2,12/01/2010 08:26,7.65,17850.0,United Kingdom
6,536365,21730,GLASS STAR FROSTED T-LIGHT HOLDER,6,12/01/2010 08:26,4.25,17850.0,United Kingdom
7,536366,22633,HAND WARMER UNION JACK,6,12/01/2010 08:28,1.85,17850.0,United Kingdom
8,536366,22632,HAND WARMER RED POLKA DOT,6,12/01/2010 08:28,1.85,17850.0,United Kingdom
9,536367,84879,ASSORTED COLOUR BIRD ORNAMENT,32,12/01/2010 08:34,1.69,13047.0,United Kingdom


## 2. Missing Value Analysis

### Why we only drop rows where CustomerID is null:
- **CustomerID is essential** for customer-level analysis (CLV, segmentation, RFM)
- **Description nulls** can often be inferred from StockCode
- Dropping all nulls blindly loses valuable transaction data

In [5]:
# Analyze missing values
null_summary = pd.DataFrame({
    'Null Count': df_raw.isnull().sum(),
    'Null %': (df_raw.isnull().sum() / len(df_raw) * 100).round(2)
})
null_summary = null_summary[null_summary['Null Count'] > 0]

print("=== Missing Value Summary ===")
print(null_summary)
print(f"\nTotal rows with any null: {df_raw.isnull().any(axis=1).sum():,}")

=== Missing Value Summary ===
             Null Count  Null %
Description        1454    0.27
CustomerID       135080   24.93



Total rows with any null: 135,080


In [6]:
# Examine rows with null CustomerID
null_customer_sample = df_raw[df_raw['CustomerID'].isnull()].head(10)
print("Sample of transactions with missing CustomerID:")
null_customer_sample

Sample of transactions with missing CustomerID:


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/01/2010 11:52,0.00,NaN,United Kingdom
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,12/01/2010 14:32,2.51,NaN,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,12/01/2010 14:32,2.51,NaN,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,12/01/2010 14:32,0.85,NaN,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,12/01/2010 14:32,1.66,NaN,United Kingdom
1447,536544,21790,VINTAGE SNAP CARDS,9,12/01/2010 14:32,1.66,NaN,United Kingdom
1448,536544,21791,VINTAGE HEADS AND TAILS CARD GAME,2,12/01/2010 14:32,2.51,NaN,United Kingdom
1449,536544,21801,CHRISTMAS TREE DECORATION WITH BELL,10,12/01/2010 14:32,0.43,NaN,United Kingdom
1450,536544,21802,CHRISTMAS TREE HEART DECORATION,9,12/01/2010 14:32,0.43,NaN,United Kingdom
1451,536544,21803,CHRISTMAS TREE STAR DECORATION,11,12/01/2010 14:32,0.43,NaN,United Kingdom


In [7]:
# Drop only rows where CustomerID is null (required for customer analysis)
df = df_raw.dropna(subset=['CustomerID']).copy()

print(f"Rows before: {len(df_raw):,}")
print(f"Rows after dropping null CustomerID: {len(df):,}")
print(f"Rows removed: {len(df_raw) - len(df):,} ({(len(df_raw) - len(df)) / len(df_raw) * 100:.1f}%)")

Rows before: 541,909
Rows after dropping null CustomerID: 406,829
Rows removed: 135,080 (24.9%)


## 3. Remove Duplicates

In [8]:
# Check for duplicates
duplicate_count = df.duplicated().sum()
print(f"Duplicate rows found: {duplicate_count:,}")

# Remove duplicates
df = df.drop_duplicates()
print(f"Rows after deduplication: {len(df):,}")

Duplicate rows found: 5,225


Rows after deduplication: 401,604


## 4. Data Type Corrections

In [9]:
# Convert InvoiceDate to datetime
# Format: MM/DD/YYYY HH:MM (US format)
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'], format='%m/%d/%Y %H:%M')

# Convert CustomerID from float to integer
# (Was float due to NaN values in original data)
df['CustomerID'] = df['CustomerID'].astype(int)

print("Data types after conversion:")
print(df.dtypes)

Data types after conversion:
InvoiceNo                 str
StockCode                 str
Description               str
Quantity                int64
InvoiceDate    datetime64[us]
UnitPrice             float64
CustomerID              int64
Country                   str
dtype: object


## 5. Understand Invoice Patterns

### InvoiceNo Convention
- **Numeric only** (e.g., 536365): Standard sale
- **Starts with 'C'** (e.g., C536379): **Cancellation/Return** - negative quantity

In [10]:
# Analyze invoice patterns
df['IsCancellation'] = df['InvoiceNo'].astype(str).str.startswith('C')

cancellation_summary = df.groupby('IsCancellation').agg({
    'InvoiceNo': 'count',
    'Quantity': ['sum', 'mean']
}).round(2)
cancellation_summary.columns = ['Transaction Count', 'Total Quantity', 'Avg Quantity']

print("=== Invoice Type Analysis ===")
print("IsCancellation = True means invoice starts with 'C' (return/cancellation)")
print(cancellation_summary)

# Verify cancellations have negative quantities
cancellations = df[df['IsCancellation']]
print(f"\nCancellations with negative quantity: {(cancellations['Quantity'] < 0).sum():,} / {len(cancellations):,}")

=== Invoice Type Analysis ===
IsCancellation = True means invoice starts with 'C' (return/cancellation)
                Transaction Count  Total Quantity  Avg Quantity
IsCancellation                                                 
False                      392732         5165886         13.15
True                         8872         -273035        -30.77

Cancellations with negative quantity: 8,872 / 8,872


## 6. Filter Invalid StockCodes

### StockCode Pattern Analysis
- **Valid pattern**: `^\d{5}[a-zA-Z]*$` 
  - 5 digits, optionally followed by 1+ letters
  - Examples: 85123, 85123A, 84029G
- **Invalid codes to remove**:
  - `POST` - Postage charges
  - `D` - Discount
  - `M` - Manual adjustment
  - `C2` - Carriage
  - `DOT` - Dotcom postage
  - `BANK CHARGES` - Bank fees

### Regex Explanation: `^\d{5}[a-zA-Z]*$`
- `^` - Start of string
- `\d{5}` - Exactly 5 digits (0-9)
- `[a-zA-Z]*` - Zero or more letters
- `$` - End of string

In [11]:
# Identify invalid StockCodes
VALID_STOCKCODE_PATTERN = r'^\d{5}[a-zA-Z]*$'

invalid_stockcodes = df[~df['StockCode'].astype(str).str.match(VALID_STOCKCODE_PATTERN)]

print("=== Invalid StockCode Analysis ===")
print(f"Invalid records: {len(invalid_stockcodes):,} ({len(invalid_stockcodes)/len(df)*100:.2f}%)")
print("\nInvalid StockCode breakdown:")
print(invalid_stockcodes.groupby(['StockCode', 'Description']).size().sort_values(ascending=False).head(10))

=== Invalid StockCode Analysis ===
Invalid records: 1,915 (0.48%)

Invalid StockCode breakdown:
StockCode     Description               
POST          POSTAGE                       1196
M             Manual                         460
C2            CARRIAGE                       134
D             Discount                        77
DOT           DOTCOM POSTAGE                  16
CRUK          CRUK Commission                 16
BANK CHARGES  Bank Charges                    12
PADS          PADS TO MATCH ALL CUSHIONS       4
dtype: int64


In [12]:
# Remove invalid StockCodes (non-product entries)
rows_before = len(df)
df = df[df['StockCode'].astype(str).str.match(VALID_STOCKCODE_PATTERN)]

print(f"Rows removed (invalid StockCode): {rows_before - len(df):,}")
print(f"Rows remaining: {len(df):,}")

Rows removed (invalid StockCode): 1,915
Rows remaining: 399,689


## 7. Final Validation

In [13]:
# Drop the temporary IsCancellation column (will recreate in feature engineering if needed)
df = df.drop(columns=['IsCancellation'])

# Final data quality check
print("=== Final Data Quality Report ===")
print(f"\nShape: {df.shape[0]:,} rows x {df.shape[1]} columns")
print(f"\nDate range: {df['InvoiceDate'].min()} to {df['InvoiceDate'].max()}")
print(f"\nUnique values:")
print(f"  - Customers: {df['CustomerID'].nunique():,}")
print(f"  - Products: {df['StockCode'].nunique():,}")
print(f"  - Invoices: {df['InvoiceNo'].nunique():,}")
print(f"  - Countries: {df['Country'].nunique()}")

print(f"\nData types:")
print(df.dtypes)

=== Final Data Quality Report ===

Shape: 399,689 rows x 8 columns

Date range: 2010-12-01 08:26:00 to 2011-12-09 12:50:00

Unique values:
  - Customers: 4,363
  - Products: 3,676
  - Invoices: 21,788
  - Countries: 37

Data types:
InvoiceNo                 str
StockCode                 str
Description               str
Quantity                int64
InvoiceDate    datetime64[us]
UnitPrice             float64
CustomerID              int64
Country                   str
dtype: object


In [14]:
# Verify no remaining nulls in critical columns
print("\nNull check (should all be 0):")
print(df.isnull().sum())


Null check (should all be 0):
InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64


In [15]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom


## 8. Export Cleaned Data

In [16]:
# Save cleaned data
df.to_csv('clean_data.csv', index=False)

print(f"Cleaned data exported to clean_data.csv")
print(f"\n=== Cleaning Summary ===")
print(f"Original rows: {len(df_raw):,}")
print(f"Final rows: {len(df):,}")
print(f"Rows removed: {len(df_raw) - len(df):,} ({(len(df_raw) - len(df)) / len(df_raw) * 100:.1f}%)")
print(f"\nReasons for removal:")
print(f"  - Null CustomerID: ~135K rows")
print(f"  - Duplicates: ~5K rows")
print(f"  - Invalid StockCodes: ~2K rows")

Cleaned data exported to clean_data.csv

=== Cleaning Summary ===
Original rows: 541,909
Final rows: 399,689
Rows removed: 142,220 (26.2%)

Reasons for removal:
  - Null CustomerID: ~135K rows
  - Duplicates: ~5K rows
  - Invalid StockCodes: ~2K rows
